File: Tp3/Punto3/ej3.py
#!/usr/bin/env python3
from math import sqrt
def test(lista:list):
    suma=0
    for i in lista:
        try: #esto esa asi, si tiene la posicion 2 de la terna no tira error, si tira erro es por q no está entonces es una pokeparada
            suma-=i[2]
        except:
            suma+=3
    return suma
def distancias(lista:list):
    suma=0
    prev=lista[0]
    for i in lista:
        suma+=sqrt((i[0]-prev[0])**2 +(i[1]-prev[1])**2) # ** = potencia
    return suma

def tirarPokeParadas(pokeParadas:list,gimnasios:list,k:int):
    """
    Esto lo que hace es armar un nuevo array tal que tenga el orden de los gimnasios tal cual le llego pero antes de cada gimnasio
    tenes la cantidad necesaria de pokeparadas y las mas cercanas a ese gimnasio, luego son descartadas y si hay un gimnasio con esa pokeparada q le quede mas convenientemente,
    no la va a tomar en cuenta
    """
    suma=0
    res=[]
    for gym in gimnasios:
        while(suma<gym[2]):
            if len(pokeParadas) <1:()
                return -1
            suma+=3
            if(suma>k):
                suma=k
            pokeParadas=sorted(pokeParadas, key=lambda pokeParada: -sqrt((gym[0]-pokeParada[0])**2 +(gym[1]-pokeParada[1])**2))
            res.append(pokeParadas.pop())
        res.append(gym)
        suma-=gym[2]
    for p in pokeParadas:
        res.append(p)
    #return res,
    return distancias(res)

def copiaSwapeada(lista,pos_uno, pos_dos):
    copia = list(lista)
    copia[pos_uno] , copia[pos_dos] =lista[pos_dos],lista[pos_uno]
    return copia
def vecinitud (lista):
    n = len(lista)
    return [(i,j) for i in range(n) for j in range(n) if j!=i]
#main   X
n,m,k= tuple(map(int, input().split())) #toma n m y k como integers
gimnasios=[tuple(map(int, input().split())) for _ in range(n)] #toma los gimnasios como ternas de ints
pokeParadas=[tuple(map(int, input().split())) for _ in range(m)] #toma las pokeparadas como tuplas de its
#gimnasios=sorted(gimnasios, key= lambda gimnasio: gimnasio[2]) #ordena por el peso del gimnasio

tuplasVecinos= [(i,j) for i in range(n) for j in range(n) if j!=i]
actual = (gimnasios, tirarPokeParadas(pokeParadas, gimnasios, k))
mejora = True
while (mejora):
    vecindad = [(copiaSwapeada(actual[0], tupla[0], tupla[1]), tirarPokeParadas(pokeParadas,copiaSwapeada(actual[0], tupla[0], tupla[1]),k)) for tupla in tuplasVecinos ]
    vecindad = [(i,j) for i,j in vecindad if j!=-1]
    minim = min(vecindad, key=lambda res: res[1])
    mejora= minim[1]< actual[1]
    if(mejora):
        actual=minim
print(actual)

#print("Swapeo: "+str(tupla)+"=> ",tirarPokeParadas(pokeParadas,gimnasiosAA,k))
"""
5 9 10
0 1 5
0 3 3
0 6 3
1 7 10
20 20 5
1 1
7 2
0 0
0 2
0 4
0 5
7 7
9 1
200 200
"""

File: Tp3/Punto3/ej3.cpp
#include "ej3.h"

double tirarPokeParadas(vnod pokeParadas, vnod &gimnasios, Mochila mochila) {
    vnod solucion;
    return tirarPokeParadas(pokeParadas, gimnasios, mochila, solucion);
}
vnod copiaSwapeada(vnod gimnasios, int pos_uno, int pos_dos) {
    vnod copia = gimnasios;
    iter_swap(copia.begin() + pos_uno, copia.begin() + pos_dos);
    return copia;
}
vvnod vecindad(vnod gimnasios) {
    vvnod res;
    for (size_t i = 0; i < gimnasios.size(); i++) {
        for (size_t j = 0; j < gimnasios.size(); j++) {
            if (i == j)
            continue;
            res.push_back(copiaSwapeada(gimnasios, i, j));
        }
    }
    return res;
}
tuple<vnod, double> solucionesVecindad(vvnod &vecindad, vnod pokeParadas, Mochila &mochila, double valor) {
    tuple<vnod, double> res;
    vnod vacio;
    res = make_tuple(vacio, valor + 1);
    for (auto gimnasios : vecindad) {
        auto sol = tirarPokeParadas(pokeParadas, gimnasios, mochila);
        if (sol == -1)
        continue;
        //vint &solucion
        if (sol < valor) {
            res = make_tuple(gimnasios, sol);
        }
    }
    return res;
}

vnod dameGimnasios(vnod gimnasios, vint &solucion){
    vnod res;//(gimnasios.size());
    int indiceAPoner=0;
    for (auto indice : solucion){
        for (auto gym: gimnasios){
            if(gym.DameIndice() == indice){
                //res[indiceAPoner]=gym;
                res.push_back(gym);
                indiceAPoner++;
                break;
            }
        }
    }
    return res;
}
tuple<vnod, double> actualInicial(vnod gimnasios, vnod pokeParadas, Mochila &mochila){
	auto ini = tirarPokeParadas(pokeParadas, gimnasios, mochila);
    return make_tuple(gimnasios, ini);
}
double mejorarSolucion (vnod gimnasios, vnod pokeParadas, Mochila &mochila, vint &solucion) {
    bool mejora = true;

    auto actual =actualInicial(gimnasios, pokeParadas, mochila);

    while (mejora) {
        vvnod vecinitud = vecindad(gimnasios);
        auto minim =
        solucionesVecindad(vecinitud, pokeParadas, mochila, get<1>(actual));
        mejora = get<1>(minim) < get<1>(actual);
        if (mejora)
        actual = minim;
    }

    vnod solucioNodos;
    solucioNodos.push_back(Nodo(27, 27, 27, 27));
    auto res = tirarPokeParadas(pokeParadas, gimnasios, mochila, solucioNodos);
    for (auto nodo : solucioNodos) {
        solucion.push_back(nodo.DameIndice());
    }

    return res;
}
double pto3(vnod gimnasios, vnod pokeParadas, Mochila mochila, vint &solucion) {
	//cerr<<"Todos putos"<<endl;
    sort(gimnasios.begin(), gimnasios.end(),[](Nodo a, Nodo b){return a.DamePociones()<b.DamePociones();});
    return mejorarSolucion(gimnasios,pokeParadas, mochila, solucion);
}

double BusquedaLocal(vnod &gimnasios, vnod &pokeParadas, Mochila &mochila, vint &solucion, double valor){
    vnod nuevosGimnasios = dameGimnasios(gimnasios,solucion);
    vint solucion_posible;
    double res_posible;
    res_posible= mejorarSolucion( nuevosGimnasios, pokeParadas,mochila,  solucion_posible);
    if(res_posible<valor){
        solucion=solucion_posible;
        return res_posible;
    }else{
        return valor;
    }
}
